napari/layers/vectors/vectors.py

```diff
@@ -653,10 +653,14 @@ def _slice_data(
             values, based on how far from the current slice they originate.
         """
         not_disp = list(self._dims_not_displayed)
-        indices = np.array(dims_indices)
+        # We want a numpy array so we can use fancy indexing with the non-displayed
+        # indices, but as dims_indices can (and often/always does) contain slice
+        # objects, the array has dtype=object which is then very slow for the
+        # arithmetic below.
+        not_disp_indices = np.array(dims_indices)[not_disp].astype(float)
         if len(self.data) > 0:
             data = self.data[:, 0, not_disp]
-            distances = abs(data - indices[not_disp])
+            distances = abs(data - not_disp_indices)
```

Comment:
Though the structure is a little different, this code block is almost identical to the one in `Points._slice_data`. The only major difference is here we calculate `projected_lengths` instead of `sizes`, but from there it looks like the logic is the same once you swap `alpha_match` for `size_match`.

Ideally, I'd like this method to call the same function. That way we don't need to keep the code and comments in sync.

Reply:
Agree! On the other hand, I think it could definitely be done in a separate PR as to not block this improvement.

Reply:
I briefly looked at this, but there are some bits which aren't 1:1 that will take some thought to work out. It could be done, but I think leaving it to a separate PR is best.